shadowsocks/dialer.go

```diff
@@ -0,0 +1,32 @@
+package shadowsocks
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+type LocalDialer struct {
+	connections chan net.Conn
+}
+
+func (d *LocalDialer) DialStream(ctx context.Context, addr string) (transport.StreamConn, error) {
+	cliConn, ok := ctx.Value(ClientConnCtxKey{}).(transport.StreamConn)
+	if !ok {
+		return nil, fmt.Errorf("expected stream connection in context but received type %T", ctx.Value(ClientConnCtxKey{}))
+	}
+
+	c1, c2 := net.Pipe()
+	a := &tcpConnAdapter{c1}
+	b := &lfwd{
+		Conn:           c2,
+		remoteAddr:     cliConn.RemoteAddr(),
+		clientTCPConn:  cliConn,
+		upstreamTarget: addr,
+	}
+	d.connections <- b
```

Comment:
Given how difficult it was for both of us to understand this area of the code, could you add a comment explaining what's going on here?  I think something like, "We treat the Lantern proxy as the upstream, effectively proxying from the client to the actual proxy code."  But feel free to use your own words of course!